### ISSUE TITLE
how to load image without actual files on server?

### ISSUE THREAD

**Issue by aleeex325** [openseadragon/openseadragon#1699]
Hi, I have some questions about loading image with OSD.

Currently, I have a whole slide image file on my local computer with a program to generate tiles based on level, x and y. The local computer is connected to a server.

I want to show the slide on a remote browser with OSD. However, there’s no tile file on the server. They will be generated when needed. Once the tile data is generated, it will be sent to browser (base64 string) via websockets, which means there’s no actual tile file during the whole process.

Process
1. remote -> (level, x, y) -> server -> (level, x, y) -> local
2. local generates tile based on level, x, y
3. local -> tile image data(base64) -> server -> tile data(base64) -> remote 
4. remote show tile image with OSD

Q1: How can I use OSD to load image files in above situation?
Q2: Which part of OSD code should I modify to implement a customized tile loading process?

Looking forward to someone who can help me with the questions or just give me some advice on building a remote viewing system.

Thanks!

**Comment by iangilman** [comment]
Sounds like an interesting project! I think you could make a custom TileSource and have its `getTileUrl` function return data URIs based on the data you get from the WebSocket. One challenge there is `getTileUrl` is not asynchronous, so it would expect you to already have the data. One possible way around that would be to give it a blank data URI (preferably of the correct dimensions) and then fill in the data once you have it via the "pre-draw" mechanism: https://github.com/openseadragon/openseadragon/pull/348

BTW, you should also check out #119 for further discussion in this area. I still think it would be great to have more robust support for non-file-based images!

**Comment by aleeex325** [comment]
@iangilman Thanks a lot. I'll try to make it work.

**Comment by JElimpus** [comment]
Hey, First time posting but thought i could give you a little help.

I use OSD using a remote tileset using base64 and AJAX
It looks something similar to this


 $.ajax({
        type: "POST",
        url: "Webservices/Map",
        data: "{TilesetID: '" + Tid + "'}",
        contentType: "application/json; charset=utf-8",
        dataType: "json",
        success: function (response) {
            MapTileSet = response.d;
            source = { type: "image", url: "data:image/png;base64," + MapTileSet.Image, height: MapTileSet.height, width: MapTileSet.width};
        },
        failure: function (response) {
            alert(response.d);
        }, complete: function () {
            if (typeof (viewer) !== "undefined") {
                viewer.destroy();
            }
            viewer = OpenSeadragon({
                id: "content_div",
                visibilityRatio: 1,
                showNavigator: true,
                sequenceMode: true,
                prefixUrl: "js/images/",
                maxZoomLevel: 10,
                defaultZoomLevel: 0.8,
                minZoomLevel: 0.8,
                tileSources: source
            });
            viewer.zoomPerClick = 1;
        }
    });

You could always save various base64 objects in an array or like i do as part of a database and recall them using AJAX.

I use this method for various things and it seems pretty full proof the good thing is you could even change the tiles based on zoom with abit more thinking with parameters sent to the server.

Sorry if i misunderstood the question but hope i helped.

**Comment by aleeex325** [comment]
@JElimpus Hi, thanks for your suggestion. It's a good idea. I'll try to understand it and integrate it into my project to see what happens. Once I get some results, we can discuss more about this solution. You definitely help me a lot. Thanks again!

**Comment by aleeex325** [comment]
@iangilman Hi, I'm wondering whether I can get a tile object by (level,x,y) and change the tiledimage for this tile?

**Comment by iangilman** [comment]
Well, if you provide your own tile source with its own `getTileUrl` you can have it return whatever kind of URL you want. Is that what you mean?

**Comment by chintuuuu** [comment]
when ever loading .svs file without connect to server
from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, https.

**Comment by iangilman** [comment]
@chintuuuu That doesn't sound like it's related to this issue. You should file a new issue for it.

**Comment by chintuuuu** [comment]
why .svs file convert to .DZI
how load .svs file

**Comment by iangilman** [comment]
@chintuuuu .svs isn't a format browsers understand. You need to convert it somehow into a format the browser can display. It doesn't need to be DZI, but it needs to be something. DZI is composed of tiles that are either JPEG or PNG, both of which are formats browsers can display. If the images are small enough, you can just convert them directly into JPEG or PNG and skip the DZI step.

**Comment by KempWatson** [comment]
@aleeex325 @chintuuuu We've actually been doing exactly this for over ten years, and indeed with SVS files as well as with all other pathology formats. There is no need for the backend to have physical "tile files", only to be able to generate them on demand from your source files, in this case an SVS or other content.

@iangilman is correct, a custom TileSource is required (unless you transcode to DeepZoom, Zoomify, IIIF etc), but to be clear you also need a backend to "make" the tiles for you on demand, from the source, in a format that matches the request wire protocol required by the TileSource. For us, the OSD front-end effort was dwarfed by the backend coding required for this.

@chintuuuu take a peek at https://the.pathology network and you will see a bog-standard OSD environment loading SVS and other files.

**Comment by zhusihan-python** [comment]
> source = { type: "image", url: "data:image/png;base64,

@JElimpus hello, i cann't see how to attach the z(level), x, y in your ajax request, what i mean is pass the x, y, z from openseadragon to your request, i want to use qwebchannel to request the tiles, hope you can see this after 4 years
